### PR TITLE
Fix regex for Topic entity

### PR DESCRIPTION
**Pull request by antoniyatrifonova**
According the specification (https://www.eclipse.org/ditto/protocol-specification-topic.html#messages-criterion-actions) for messages creation, the action segment can be freely chosen.

Signed-off-by: Antonia Trifonova antonia.trifonova@bosch.io

### CODE REVIEW

**Comment by antoniyatrifonova** [comment]
@konstantina-gramatova please review this PR

File: protocol/topic.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ const (
 )
 
 var regexFiveElementTopic = regexp.MustCompile("^([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)$")
-var regexSixElementTopic = regexp.MustCompile("^([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)$")
+var regexSixElementTopic = regexp.MustCompile("^([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)/(.*)$")
```
Comment (konstantina-gramatova) [inline protocol/topic.go:92]: Can we simplify the regexp. here and replace the two separate ones with only one - something like e.g. `"^([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)(/([^/]{1}.+))?$"` 
?
It's not a strict required RFC matcher but it's a start :)
That way only one regexp will be compiled and useful in both cases as well as helpful to simplify the logic using it.

File: protocol/topic.go [hunk 2]
```diff
@@ -157,7 +157,12 @@ func (topic *Topic) UnmarshalJSON(data []byte) error {
...
-		topic.Action = TopicAction(elements[index])
+		if topic.Criterion == CriterionMessages {
+			i := strings.Index(v, elements[index])
+			topic.Action = TopicAction(v[i:])
+		} else {
+			topic.Action = TopicAction(elements[index])
+		}
 	} else {
 		topic.Action = ""
 	}
```
Comment (konstantina-gramatova) [inline protocol/topic.go:168]: If only one regexp. is used - the logic here can be simplified (i.e. omit manual string splitting and concatenating) and rely on the matching groups in the regexp.
Please, see - func (*Regexp) [FindAllStringSubmatch](https://cs.opensource.google/go/go/+/go1.17.7:src/regexp/regexp.go;l=1190)

**Review by konstantina-gramatova — CHANGES_REQUESTED** [review]

File: protocol/topic.go [hunk 3]
```diff
@@ -154,14 +157,16 @@ func (topic *Topic) UnmarshalJSON(data []byte) error {
...
 	topic.Criterion = TopicCriterion(elements[index])
 	index++
-	if index < len(elements) {
-		topic.Action = TopicAction(elements[index])
-	} else {
-		topic.Action = ""
+
+	if strings.HasPrefix(elements[index], "/") {
+		index++
 	}
```
Comment (konstantina-gramatova) [inline protocol/topic.go:166]: Do we really need this check? The size of `elements` will always be a fixed one and the last element will always be a "" if there's no match for it?

**Review by konstantina-gramatova — APPROVED** [review]
LGTM

**Pull request merged**